### PR TITLE
CDAP-3189 Avoid adding cdap-default.xml multiple times to the configuration.

### PR DESCRIPTION
**Pull request by sagarkapare**
JIRA: https://issues.cask.co/browse/CDAP-3189

In order to pass the cdap configurations (cdap-default.xml and cdap-site.xml) to the mapper and reducer tasks, they are serialized into the hConf under property name "hconf.cconf".

While deserializing, we first create the new `CConfiguration` object which includes the cdap-default.xml and cdap-site.xml. Then to this object we add the value of "hconf.cconf" property as a new resource. This results in the `CConfiguration` object to contain duplicate configurations. 

As a result we see lot of warning messages when the MapReduce program is run:
```java
2015-09-11 09:23:00,407 - WARN  [LocalJobRunner Map Task Executor #5:c.c.c.c.c.Configuration@1748] - java.io.ByteArrayInputStream@6158ab02:an attempt to override final parameter: app.template.plugin.dir;  Ignoring.
2015-09-11 09:23:00,407 - WARN  [LocalJobRunner Map Task Executor #4:c.c.c.c.c.Configuration@1748] - java.io.ByteArrayInputStream@11b48062:an attempt to override final parameter: app.template.plugin.dir;  Ignoring.
2015-09-11 09:23:00,409 - WARN  [LocalJobRunner Map Task Executor #2:c.c.c.c.c.Configuration@1748] - java.io.ByteArrayInputStream@5f8ce5e:an attempt to override final parameter: stream.instance.file.prefix;  Ignoring.
2015-09-11 09:23:00,409 - WARN  [LocalJobRunner Map Task Executor #5:c.c.c.c.c.Configuration@1748] - java.io.ByteArrayInputStream@6158ab02:an attempt to override final parameter: stream.instance.file.prefix;  Ignoring.
``` 

In this PR, while deserializing, we create an empty `CConfiguration` object which will have no resource added to it and then to this object we add the value of "hconf.cconf" property as a new resource.

 Tested on standalone and I no longer see those warning messages for `app.template.plugin.dir` and `stream.instance.file.prefix`

Will run the integration tests on cluster as well.

### CODE REVIEW

**Comment by sreevatsanraman** [comment]
Add the reference to the build from the feature branch. Great info in the PR, can this be updated to the JIRA as well.

LGTM. Please merge if the tests on cluster pass.

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-DUT2813-1

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java [hunk 1]
```diff
@@ -265,7 +265,7 @@ private void setConf(CConfiguration conf) {
   }
 
   public CConfiguration getConf() {
-    CConfiguration conf = CConfiguration.create();
+    CConfiguration conf = CConfiguration.createEmpty();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java]: There are many places we do similar thing by calling `cConf.clear()` first (search for usage of clear(), you will see a quite a number). It's better to unify to do it in one way. I propose to do it like this:

```java
CConfiguration cConf = CConfiguration.create(resource);

...
// Two new factory method to create with the give resources only
public static CConfiguration create(URI resource, URI...moreResources);
public static CConfiguration create(InputStream resource);
```

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java [hunk 2]
```diff
@@ -265,7 +265,7 @@ private void setConf(CConfiguration conf) {
...
     conf.addResource(new ByteArrayInputStream(hConf.get(HCONF_ATTR_CCONF).getBytes()));
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java]: The existing code is actually wrong. Should never call `String.getBytes()` without providing encoding. Better use `Bytes.toBytes(hConf.get(HCONF_ATTR_CCONF))`

Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java]: Configuration has a method ``addResource(String)``, can we just use that instead ?

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java]: You can creates as many static create(X) method :)

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java (hunk 1 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java]: Added static methods. Using URL as a parameter for the create to make it in consistent with the `addResource` method parameters. Please let me know if this is ok.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java (hunk 2 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java]: Fixed.

**Comment by sagarkapare** [comment]
@chtyim Addressed comments. Please take a look again.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java [hunk 3]
```diff
@@ -193,9 +193,7 @@ public void initialize(TwillContext context) {
 
       UserGroupInformation.setConfiguration(hConf);
 
-      cConf = CConfiguration.create();
-      cConf.clear();
-      cConf.addResource(new File(configs.get("cConf")).toURI().toURL());
+      cConf = CConfiguration.create(new File(configs.get("cConf")).toURI().toURL());
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java]: Maybe having a create method that takes File.

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java]: more precisely, `create(File file, File...moreFiles)`.

**Comment by chtyim** [comment]
Just one comment left.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java]: done.

**Comment by sagarkapare** [comment]
@chtyim Addressed comments. Please take a look again. Thanks!

**Comment by chtyim** [comment]
LGTM

**Comment by sagarkapare** [comment]
The build passed here - http://builds.cask.co/browse/CDAP-DUT2813-3. Also tested on the cluster with the help of integration tests. No log statements  `an attempt to override final parameter: app.template.plugin.dir;  Ignoring.` and `an attempt to override final parameter: stream.instance.file.prefix;  Ignoring.` observed in the master log files and MapReduce completed. Merging.

**Pull request merged**